File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandBuilders/DataCommandBuilder.java
package bgu.spl171.net.impl.TFTP.CommandBuilders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.Data;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;

public class DataCommandBuilder implements CommandBuilder {
    Short size = null;
    Short blockId = null;
    private byte[] data;
    private int index = 0;
    private final ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();

    @Override
    public Command decodeNextByte(byte nextByte) {
        if (size == null) {
            size = shortEncoderDecoder.decodeNextByte(nextByte);
            data = new byte[size];
        } else if (blockId == null) {
            blockId = shortEncoderDecoder.decodeNextByte(nextByte);
        } else {
            data[index] = nextByte;
            if (++index == size) {
                return new Data(size, blockId, data);
            }
        }
        return null;
    }
}

File: java to c++ builders/DataCommandBuilder.cpp
#include "DataCommandBuilder.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandBuilders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::Data;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;

						Command *DataCommandBuilder::decodeNextByte(char nextByte)
						{
							if (!size)
							{
								size = shortEncoderDecoder->decodeNextByte(nextByte);
								data = std::vector<char>(size);
							}
							else if (!blockId)
							{
								blockId = shortEncoderDecoder->decodeNextByte(nextByte);
							}
							else
							{
								data[index] = nextByte;
								if (++index == size)
								{
									return new Data(size, blockId, data);
								}
							}
							return nullptr;
						}
					}
				}
			}
		}
	}
}
